File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/nms/listener/NmsPacketListener.java
package dev.slne.surf.surfapi.bukkit.api.nms.listener;

import com.google.common.reflect.TypeToken;
import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution;
import dev.slne.surf.surfapi.bukkit.api.nms.listener.packets.NmsPacket;
import dev.slne.surf.surfapi.shared.api.util.InternalSurfApi;
import org.jetbrains.annotations.ApiStatus;
import org.jspecify.annotations.NullMarked;

@NmsUseWithCaution
@NullMarked
public interface NmsPacketListener<Packet extends NmsPacket> {

    @ApiStatus.Internal
    @InternalSurfApi
    default Class<? super Packet> getPacketClass() {
        TypeToken<Packet> typeToken = new TypeToken<Packet>(getClass()) {
        };

        return typeToken.getRawType();
    }
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/nms/listener/NmsPacketListener.kt
package dev.slne.surf.surfapi.bukkit.api.nms.listener

import com.google.common.reflect.TypeToken
import dev.slne.surf.surfapi.bukkit.api.nms.NmsUseWithCaution
import dev.slne.surf.surfapi.bukkit.api.nms.listener.packets.NmsPacket

@NmsUseWithCaution
sealed interface NmsPacketListener<Packet : NmsPacket> {
    val packetClass: Class<in Packet>
        get() {
            val typeToken = object : TypeToken<Packet>(javaClass) {}
            return typeToken.rawType
        }
}
